backend/src/v2/driver/driver.go

```diff
@@ -1777,14 +1785,42 @@ func resolveUpstreamParameters(cfg resolveUpstreamOutputsConfig) (*structpb.Valu
 			if subTaskName == "" {
 				return nil, cfg.err(fmt.Errorf("producer_subtask not in outputParams"))
 			}
-			glog.V(4).Infof(
-				"Overriding currentTask, %v, output with currentTask's producer_subtask, %v, output.",
-				currentTask.TaskName(),
-				subTaskName,
-			)
-			currentTask, ok = tasks[subTaskName]
-			if !ok {
-				return nil, cfg.err(fmt.Errorf("subTaskName, %v, not in tasks", subTaskName))
+			// If the sub-task is a ParallelFor DAG, we need to gather all i.e. dsl.Collected support
+			if currentTask.GetExecution().GetCustomProperties()["iteration_count"] != nil {
+				glog.V(4).Infof("currentTask, %v, is a ParallelFor DAG. Attempting to gather all producer_subtask iterations of %w", currentTask.TaskName(), subTaskName)
+
+				outputs_list := make([]*structpb.Value, 0)
+				for i := range currentTask.GetExecution().GetCustomProperties()["iteration_count"].GetIntValue() {
+					// Follow the convention set in mlmd.GetExecutionsInDAG for ParallelFor tasks
+					subTaskIterationName := fmt.Sprintf("%s_idx_%v", subTaskName, i)
+					glog.V(4).Infof("subTaskIterationName: %v", subTaskIterationName)
+					subTask, ok := tasks[subTaskIterationName]
+					if !ok {
+						return nil, cfg.err(fmt.Errorf("subTaskName, %v, not in tasks", subTaskName))
+					}
+					_, outputParametersCustomProperty, err := subTask.GetParameters()
+					if err != nil {
+						return nil, err
+					}
+					outputs_list = append(outputs_list, outputParametersCustomProperty[outputParameterKey])
+				}
+				return &structpb.Value{
+					Kind: &structpb.Value_ListValue{
+						ListValue: &structpb.ListValue{
+							Values: outputs_list,
+						},
+					},
+				}, nil
+			} else {
```

Comment:
Nit: You don't need the else since you always return in the `if`. It helps with readability in my opinion to fix this.